design-extension-notes.md

```diff
@@ -0,0 +1,149 @@
+
+- Secure processors: Designing and integrating enclaves is a large design effort, with significant challenges. In particular, enclave environments have significant resource restrictions and limited functionalitie. Some notes on our approach follow:
+    - We intend to incorporate secure processors on the key server(s). Our end goal is that the [system's cryptographic functionalities](system-functionalities.md) and OPAQUE handshake are run in an enclave. Other non-core functionalities can be handled outside of the enclave.
+    - We will design an enclave abstraction and define a path for future integration of multiple enclave technologies, starting with SGX and AWS Nitro. 
+    - We will start with an exploration of appropriate enclave boundaries. The goal is to use secure processors for a minimal set of functionalities, e.g., for all server-side computations on sensitive user data, but excluding, to the extent possible, session handling, message framework, general user management.
+- MPC: We intend to incorporate multiple key servers, each of which contributes to key generation and use. 
+    - In this way, we can distribute trust across multiple cloud providers, with each provider having limited access to system and user data. In this setting, we have `n` servers `S_1, ..., S_n` and a threshold parameter `t`.
+    - Architecturally, it may also be possible to use an aggregator to help with routing and processing of messages from the key servers on behalf of the client, as well as in the final processing of a signature request (i.e., the formation of a signature from partial signatures, where the partial signatures are provided by a subset of key servers).
+
+
+## Requirements and High-level Design Notes
+### Threat Model 
+See the [Design Philosophy and Threat Model section](design-philosophy-and-threat-model) for information on our existing approach and requirements with respect to threat model. Of particular note for our system extensions:
+1. Password-guessing attacks: Our system MUST be robust against password-guessing attacks. Specifically, the system should satisfy the following properties:
+    1. A suitably robust password choice renders the probability of a successful attack negligible.
+    1. Servers limit the number of authentication attempts a user makes and actively respond to possible password-guessing attacks to prevent key theft and misuse.
+1. Key server storage: We treat databases as untrusted to the extent possible. This is in line with our principle of limiting trust placed in individual components (and therefore the attack surface) as much as possible. Additionally, this choice ideally allows for flexible deployment scenarios, such as the use of managed persistence solutions.
+
+Enclave integration notes:
+- Password guessing attacks: We can consider using hardware-encrypted RAM to track login attempts.
```

Comment:
Also, do we want to add a note about key server storage here? Like, the enclave can encrypt data under an enclave-specific key and export it to storage, and this achieves our key server storage goals.

Reply:
Yes, because I'm currently focused on the less obvious design considerations around enclave integration, not on how MPC and enclaves help us achieve our more general security goals already described in the linked section.

Also, did you look at the linked threat model section? Because these points are I believe discussed there already, although perhaps not in the level of detail we might want.

Reply:
To the second comment: there is a note about this option for key server storage in the section on Secure Storage below.

Reply:
Okay. It might help me to have a note at the top of this section to clarify that we're not exhaustively explaining how these extensions help us achieve our security goals (or maybe, not going to list out the general / common benefits of these features).

Reply:
These are just design extension notes. It's not meant to cover the threat model here.